File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.CachingProxyExample/WeatherServiceProxy.cs
namespace ProxyDesignPattern.CachingProxyExample;

/// <summary>
/// Represents a cached item with expiration time.
/// </summary>
public class CacheItem
{
    public string Value { get; }
    public DateTime ExpiresAt { get; }

    public CacheItem(string value, TimeSpan ttl)
    {
        Value = value;
        ExpiresAt = DateTime.Now.Add(ttl);
    }

    public bool IsExpired => DateTime.Now > ExpiresAt;
}

/// <summary>
/// Caching Proxy with TTL (Time To Live) - Cache expires after a specified duration.
/// </summary>
public class WeatherServiceProxy : IWeatherService
{
    private readonly RealWeatherService _realService;
    private readonly Dictionary<string, CacheItem> _cache;
    private readonly TimeSpan _cacheTtl;

    public WeatherServiceProxy(TimeSpan? cacheTtl = null)
    {
        _realService = new RealWeatherService();
        _cache = new Dictionary<string, CacheItem>(StringComparer.OrdinalIgnoreCase);
        _cacheTtl = cacheTtl ?? TimeSpan.FromMinutes(5); // Default: 5 minutes
        
        Console.WriteLine($"[Proxy] Cache TTL set to: {_cacheTtl.TotalSeconds} seconds");
    }

    public string GetWeather(string city)
    {
        // Check cache first
        if (_cache.TryGetValue(city, out var cachedItem))
        {
            if (!cachedItem.IsExpired)
            {
                Console.WriteLine($"[Proxy] Cache HIT for '{city}' (expires in {(cachedItem.ExpiresAt - DateTime.Now).TotalSeconds:F0}s)");
                return cachedItem.Value;
            }
            
            // Cache expired - remove it
            Console.WriteLine($"[Proxy] Cache EXPIRED for '{city}'");
            _cache.Remove(city);
        }
        else
        {
            Console.WriteLine($"[Proxy] Cache MISS for '{city}'");
        }

        // Fetch from real service
        var result = _realService.GetWeather(city);

        // Store in cache with TTL
        _cache[city] = new CacheItem(result, _cacheTtl);
        Console.WriteLine($"[Proxy] Data cached for '{city}' (TTL: {_cacheTtl.TotalSeconds}s)");

        return result;
    }

    public void ClearCache()
    {
        _cache.Clear();
        Console.WriteLine("[Proxy] Cache cleared!");
    }

    public int CacheSize => _cache.Count;
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.CachingProxyExample/src/WeatherServiceProxy.ts
import { CacheItem } from "./CacheItem";
import { IWeatherService } from "./IWeatherService";
import { RealWeatherService } from "./RealWeatherService";

/**
 * Caching Proxy with TTL (Time To Live) - Cache expires after a specified duration.
 */
export class WeatherServiceProxy implements IWeatherService {
  private readonly _realService: RealWeatherService;
  private readonly _cache: Map<string, CacheItem>;
  private readonly _cacheTtlMs: number;

  constructor(cacheTtlMs?: number) {
    this._realService = new RealWeatherService();
    this._cache = new Map();
    this._cacheTtlMs = cacheTtlMs ?? 5 * 60 * 1000; // Default: 5 minutes

    console.log(`[Proxy] Cache TTL set to: ${this._cacheTtlMs / 1000} seconds`);
  }

  async getWeather(city: string): Promise<string> {
    const key = city.toLowerCase();

    // Check cache first
    const cachedItem = this._cache.get(key);
    if (cachedItem) {
      if (!cachedItem.isExpired) {
        const remainingSec = Math.round(
          (cachedItem.expiresAt.getTime() - Date.now()) / 1000
        );
        console.log(`[Proxy] Cache HIT for '${city}' (expires in ${remainingSec}s)`);
        return cachedItem.value;
      }

      // Cache expired - remove it
      console.log(`[Proxy] Cache EXPIRED for '${city}'`);
      this._cache.delete(key);
    } else {
      console.log(`[Proxy] Cache MISS for '${city}'`);
    }

    // Fetch from real service
    const result = await this._realService.getWeather(city);

    // Store in cache with TTL
    this._cache.set(key, new CacheItem(result, this._cacheTtlMs));
    console.log(`[Proxy] Data cached for '${city}' (TTL: ${this._cacheTtlMs / 1000}s)`);

    return result;
  }

  clearCache(): void {
    this._cache.clear();
    console.log("[Proxy] Cache cleared!");
  }

  get cacheSize(): number {
    return this._cache.size;
  }
}
